cannot use flag (type uint32) as type uintptr in argument to procRedrawWindow.Call

Hi.
I am trying to build on Windows 10.

> go\src\github.com\gonutz\d3d9\samples\fullscreen>go get -u github.com/AllenDang/gform
> # github.com/AllenDang/w32
> ..\..\..\..\AllenDang\w32\user32.go:1039:10: cannot use flag (type uint32) as type uintptr in argument to procRedrawWindow.Call

> >go env
> set GO111MODULE=
> set GOARCH=amd64
> set GOBIN=
> set GOCACHE=C:\Users\Worker\AppData\Local\go-build
> set GOENV=C:\Users\Worker\AppData\Roaming\go\env
> set GOEXE=.exe
> set GOFLAGS=
> set GOHOSTARCH=amd64
> set GOHOSTOS=windows
> set GOINSECURE=
> set GOMODCACHE=C:\Users\Worker\go\pkg\mod
> set GONOPROXY=
> set GONOSUMDB=
> set GOOS=windows
> set GOPATH=C:\Users\Worker\go
> set GOPRIVATE=
> set GOPROXY=https://proxy.golang.org,direct
> set GOROOT=c:\go
> set GOSUMDB=sum.golang.org
> set GOTMPDIR=
> set GOTOOLDIR=c:\go\pkg\tool\windows_amd64
> set GCCGO=gccgo
> set AR=ar
> set CC=gcc
> set CXX=g++
> set CGO_ENABLED=1
> set GOMOD=
> set CGO_CFLAGS=-g -O2
> set CGO_CPPFLAGS=
> set CGO_CXXFLAGS=-g -O2
> set CGO_FFLAGS=-g -O2
> set CGO_LDFLAGS=-g -O2
> set PKG_CONFIG=pkg-config
> set GOGCCFLAGS=-m64 -mthreads -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=C:\Users\Worker\AppData\Local\Temp\go-build907535614=/tmp/go-build -gno-record-gcc-switches

> >go version
> go version go1.15.2 windows/amd64

You may close this issue: I was able to launch an app, which lad to rendering in a full-screen (pink colored).
My cursor was moving - and it was under law resolution move.
I have switched to other apps (and my normal  resolution returned back) - and closed it manually.

> C:\Users\Worker>go get -u github.com/gonutz/d3d9
> 
> C:\Users\Worker>cd go
> 
> C:\Users\Worker\go>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go
> 
> 02.10.2020  12:14    <DIR>          .
> 02.10.2020  12:14    <DIR>          ..
> 02.10.2020  12:14    <DIR>          pkg
> 02.10.2020  12:13    <DIR>          src
>                0 файлов              0 байт
>                4 папок  10 373 697 536 байт свободно
> 
> C:\Users\Worker\go>cd src
> 
> C:\Users\Worker\go\src>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go\src
> 
> 02.10.2020  12:13    <DIR>          .
> 02.10.2020  12:13    <DIR>          ..
> 02.10.2020  12:13    <DIR>          github.com
>                0 файлов              0 байт
>                3 папок  10 373 697 536 байт свободно
> 
> C:\Users\Worker\go\src>cd github.com/gonutz/d3d9
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go\src\github.com\gonutz\d3d9
> 
> 02.10.2020  12:13    <DIR>          .
> 02.10.2020  12:13    <DIR>          ..
> 02.10.2020  12:13             6 645 base_texture_windows.go
> 02.10.2020  12:13             1 950 color_windows.go
> 02.10.2020  12:13            76 879 constants_windows.go
> 02.10.2020  12:13             8 759 cube_texture_windows.go
> 02.10.2020  12:13            54 482 device_windows.go
> 02.10.2020  12:13            10 065 direct3d9_windows.go
> 02.10.2020  12:13            99 029 enums_windows.go
> 02.10.2020  12:13             3 341 error_windows.go
> 02.10.2020  12:13             7 934 index_buffer_windows.go
> 02.10.2020  12:13             1 096 LICENSE
> 02.10.2020  12:13             1 844 macros_windows.go
> 02.10.2020  12:13             1 976 pixel_shader_windows.go
> 02.10.2020  12:13             2 458 query_windows.go
> 02.10.2020  12:13             6 964 README.md
> 02.10.2020  12:13             4 564 resource_windows.go
> 02.10.2020  12:13    <DIR>          samples
> 02.10.2020  12:13             1 818 state_block_windows.go
> 02.10.2020  12:13             7 086 surface_windows.go
> 02.10.2020  12:13             4 100 swap_chain_windows.go
> 02.10.2020  12:13             8 421 texture_windows.go
> 02.10.2020  12:13            15 379 types_windows.go
> 02.10.2020  12:13            12 020 vertex_buffer_windows.go
> 02.10.2020  12:13             2 063 vertex_declaration_windows.go
> 02.10.2020  12:13             2 002 vertex_shader_windows.go
> 02.10.2020  12:13             8 730 volume_texture_windows.go
> 02.10.2020  12:13             4 683 volume_windows.go
>               25 файлов        354 288 байт
>                3 папок  10 369 716 224 байт свободно
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9>go build -i -v -x
> WORK=C:\Users\Worker\AppData\Local\Temp\go-build558590870
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9>go build
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9>go test
> ?       github.com/gonutz/d3d9  [no test files]
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9>cd samples
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples
> 
> 02.10.2020  12:13    <DIR>          .
> 02.10.2020  12:13    <DIR>          ..
> 02.10.2020  12:13    <DIR>          display_info
> 02.10.2020  12:13    <DIR>          fullscreen
> 02.10.2020  12:13    <DIR>          rotating_quad
> 02.10.2020  12:13    <DIR>          simple_glow
> 02.10.2020  12:13    <DIR>          static_triangle
>                0 файлов              0 байт
>                7 папок  10 364 735 488 байт свободно
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples>cd fullscreen
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen
> 
> 02.10.2020  12:13    <DIR>          .
> 02.10.2020  12:13    <DIR>          ..
> 02.10.2020  12:13                31 build.bat
> 02.10.2020  12:13             2 532 main.go
> 02.10.2020  12:13               818 README.md
>                3 файлов          3 381 байт
>                2 папок  10 351 964 160 байт свободно
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>build.bat
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>go build -ldflags -H=windowsgui
> main.go:8:2: cannot find package "github.com/gonutz/w32" in any of:
>         c:\go\src\github.com\gonutz\w32 (from $GOROOT)
>         C:\Users\Worker\go\src\github.com\gonutz\w32 (from $GOPATH)
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>go get -u github.com/gonutz/w32
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>build.bat
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>go build -ldflags -H=windowsgui
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>dir
>  Том в устройстве C имеет метку Win10
>  Серийный номер тома: D0F6-A093
> 
>  Содержимое папки C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen
> 
> 02.10.2020  12:17    <DIR>          .
> 02.10.2020  12:17    <DIR>          ..
> 02.10.2020  12:13                31 build.bat
> 02.10.2020  12:17         2 091 520 fullscreen.exe
> 02.10.2020  12:13             2 532 main.go
> 02.10.2020  12:13               818 README.md
>                4 файлов      2 094 901 байт
>                2 папок  10 350 559 232 байт свободно
> 
> C:\Users\Worker\go\src\github.com\gonutz\d3d9\samples\fullscreen>fullscreen.exe
>

--- a/samples/display_info/main.go
+++ b/samples/display_info/main.go
@@ -19,7 +19,7 @@ func main() {
 		fmt.Printf("adapter %v has %v R8G8B8 modes:\n", adapter, modeCount)
 
 		for mode := uint(0); mode < modeCount; mode++ {
-			adapaterMode, err := d3d.EnumAdapterModes(
+			adapterMode, err := d3d.EnumAdapterModes(
 				adapter,
 				d3d9.FMT_X8R8G8B8,
 				mode,
@@ -28,9 +28,9 @@ func main() {
 			fmt.Printf(
 				"  mode %2d: %vx%v %vHz\n",
 				mode,
-				adapaterMode.Width,
-				adapaterMode.Height,
-				adapaterMode.RefreshRate,
+				adapterMode.Width,
+				adapterMode.Height,
+				adapterMode.RefreshRate,
 			)
 		}
 

--- a/samples/fullscreen/go.mod
+++ b/samples/fullscreen/go.mod
@@ -1,0 +1,10 @@
+module github.com/gonutz/d3d9/samples/fullscreen
+
+go 1.11
+
+require (
+	github.com/gonutz/d3d9 v1.2.1
+	github.com/gonutz/w32/v2 v2.2.0
+)
+
+replace github.com/gonutz/d3d9 => ../../

--- a/samples/fullscreen/main.go
+++ b/samples/fullscreen/main.go
@@ -5,7 +5,7 @@ import (
 	"syscall"
 
 	"github.com/gonutz/d3d9"
-	"github.com/gonutz/w32"
+	"github.com/gonutz/w32/v2"
 )
 
 func main() {
@@ -14,6 +14,7 @@ func main() {
 	runtime.LockOSThread()
 
 	const className = "fullscreen_window_class"
+	classNamePtr, _ := syscall.UTF16PtrFromString(className)
 	w32.RegisterClassEx(&w32.WNDCLASSEX{
 		Cursor: w32.LoadCursor(0, w32.MakeIntResource(w32.IDC_ARROW)),
 		WndProc: syscall.NewCallback(func(window w32.HWND, msg uint32, w, l uintptr) uintptr {
@@ -30,12 +31,13 @@ func main() {
 				return w32.DefWindowProc(window, msg, w, l)
 			}
 		}),
-		ClassName: syscall.StringToUTF16Ptr(className),
+		ClassName: classNamePtr,
 	})
 
+	windowNamePtr, _ := syscall.UTF16PtrFromString("Fullscreen")
 	window := w32.CreateWindow(
-		syscall.StringToUTF16Ptr(className),
-		syscall.StringToUTF16Ptr("Fullscreen"),
+		classNamePtr,
+		windowNamePtr,
 		w32.WS_OVERLAPPEDWINDOW|w32.WS_VISIBLE,
 		w32.CW_USEDEFAULT, w32.CW_USEDEFAULT, 640, 480,
 		0, 0, 0, nil,

--- a/samples/rotating_quad/main.go
+++ b/samples/rotating_quad/main.go
@@ -3,29 +3,44 @@ package main
 import (
 	"math"
 	"runtime"
+	"syscall"
 
-	"github.com/AllenDang/gform"
 	"github.com/gonutz/d3d9"
-	"github.com/gonutz/w32"
+	"github.com/gonutz/w32/v2"
 )
 
-func init() {
-	runtime.LockOSThread()
-}
-
 func main() {
-	// setup gform and create a window
-	gform.Init()
-	form := gform.NewForm(nil)
-	windowHandle := form.Handle()
-	// we want a square window interior, but the window has a title bar and
-	// borders left and right, so add some pixels to compensate for it; this is
-	// just a sample, the window content will be square enough for our purposes
-	form.SetSize(520, 540)
-	form.Show()
-	form.OnClose().Bind(func(*gform.EventArg) {
-		w32.DestroyWindow(w32.HWND(form.Handle()))
+	runtime.LockOSThread()
+
+	const className = "fullscreen_window_class"
+	classNamePtr, _ := syscall.UTF16PtrFromString(className)
+	w32.RegisterClassEx(&w32.WNDCLASSEX{
+		Cursor: w32.LoadCursor(0, w32.MakeIntResource(w32.IDC_ARROW)),
+		WndProc: syscall.NewCallback(func(window w32.HWND, msg uint32, w, l uintptr) uintptr {
+			switch msg {
+			case w32.WM_KEYDOWN:
+				if w == w32.VK_ESCAPE {
+					w32.SendMessage(window, w32.WM_CLOSE, 0, 0)
+				}
+				return 0
+			case w32.WM_DESTROY:
+				w32.PostQuitMessage(0)
+				return 0
+			default:
+				return w32.DefWindowProc(window, msg, w, l)
+			}
+		}),
+		ClassName: classNamePtr,
 	})
+
+	windowNamePtr, _ := syscall.UTF16PtrFromString("Rotating Quad")
+	windowHandle := w32.CreateWindow(
+		classNamePtr,
+		windowNamePtr,
+		w32.WS_OVERLAPPEDWINDOW|w32.WS_VISIBLE,
+		w32.CW_USEDEFAULT, w32.CW_USEDEFAULT, 640, 480,
+		0, 0, 0, nil,
+	)
 
 	d3d, err := d3d9.Create(d3d9.SDK_VERSION)
 	check(err)
@@ -87,8 +102,10 @@ func main() {
 
 	// create a timer that ticks every 10ms and register a callback for it
 	rotation := float32(0)
-	w32.SetTimer(w32.HWND(windowHandle), 1, 10, 0)
-	form.Bind(w32.WM_TIMER, func(*gform.EventArg) {
+	w32.SetTimer(windowHandle, 1, 100, 0)
+	var msg w32.MSG
+	for w32.GetMessage(&msg, 0, 0, 0) != 0 {
+		w32.TranslateMessage(&msg)
 		// create a 4x4 model-view-projection matrix that rotates the
 		// rectangle about the Z-axis
 		s, c := math.Sincos(float64(rotation))
@@ -106,9 +123,8 @@ func main() {
 		check(device.DrawPrimitive(d3d9.PT_TRIANGLESTRIP, 0, 2))
 		check(device.EndScene())
 		check(device.Present(nil, nil, 0, nil))
-	})
-
-	gform.RunMainLoop()
+		w32.DispatchMessage(&msg)
+	}
 }
 
 func check(err error) {

--- a/samples/simple_glow/main.go
+++ b/samples/simple_glow/main.go
@@ -2,25 +2,44 @@ package main
 
 import (
 	"runtime"
+	"syscall"
 
-	"github.com/AllenDang/gform"
 	"github.com/gonutz/d3d9"
-	"github.com/gonutz/w32"
+	"github.com/gonutz/w32/v2"
 )
 
-func init() {
-	runtime.LockOSThread()
-}
-
 func main() {
-	// setup gform and create a window
-	gform.Init()
-	form := gform.NewForm(nil)
-	windowHandle := form.Handle()
-	form.Show()
-	form.OnClose().Bind(func(*gform.EventArg) {
-		w32.DestroyWindow(w32.HWND(form.Handle()))
+	runtime.LockOSThread()
+
+	const className = "fullscreen_window_class"
+	classNamePtr, _ := syscall.UTF16PtrFromString(className)
+	w32.RegisterClassEx(&w32.WNDCLASSEX{
+		Cursor: w32.LoadCursor(0, w32.MakeIntResource(w32.IDC_ARROW)),
+		WndProc: syscall.NewCallback(func(window w32.HWND, msg uint32, w, l uintptr) uintptr {
+			switch msg {
+			case w32.WM_KEYDOWN:
+				if w == w32.VK_ESCAPE {
+					w32.SendMessage(window, w32.WM_CLOSE, 0, 0)
+				}
+				return 0
+			case w32.WM_DESTROY:
+				w32.PostQuitMessage(0)
+				return 0
+			default:
+				return w32.DefWindowProc(window, msg, w, l)
+			}
+		}),
+		ClassName: classNamePtr,
 	})
+
+	windowNamePtr, _ := syscall.UTF16PtrFromString("Simple Glow")
+	windowHandle := w32.CreateWindow(
+		classNamePtr,
+		windowNamePtr,
+		w32.WS_OVERLAPPEDWINDOW|w32.WS_VISIBLE,
+		w32.CW_USEDEFAULT, w32.CW_USEDEFAULT, 640, 480,
+		0, 0, 0, nil,
+	)
 
 	d3d, err := d3d9.Create(d3d9.SDK_VERSION)
 	check(err)
@@ -41,9 +60,14 @@ func main() {
 	defer device.Release()
 
 	// create a timer that ticks every 10ms and register a callback for it
-	w32.SetTimer(w32.HWND(windowHandle), 1, 10, 0)
+	w32.SetTimer(windowHandle, 1, 10, 0)
 	red, dRed := 255, -1
-	form.Bind(w32.WM_TIMER, func(*gform.EventArg) {
+
+	// This is the main message loop that runs until the window is closed.
+	var msg w32.MSG
+	for w32.GetMessage(&msg, 0, 0, 0) != 0 {
+		w32.TranslateMessage(&msg)
+
 		// clear the screen to the current color
 		check(device.Clear(
 			nil,
@@ -60,9 +84,9 @@ func main() {
 		}
 		red += dRed
 		check(device.Present(nil, nil, 0, nil))
-	})
 
-	gform.RunMainLoop()
+		w32.DispatchMessage(&msg)
+	}
 }
 
 func check(err error) {

--- a/samples/static_triangle/main.go
+++ b/samples/static_triangle/main.go
@@ -2,26 +2,44 @@ package main
 
 import (
 	"runtime"
-	"unsafe"
+	"syscall"
 
-	"github.com/AllenDang/gform"
 	"github.com/gonutz/d3d9"
-	"github.com/gonutz/w32"
+	"github.com/gonutz/w32/v2"
 )
 
-func init() {
-	runtime.LockOSThread()
-}
-
 func main() {
-	// setup gform and create a window
-	gform.Init()
-	form := gform.NewForm(nil)
-	windowHandle := unsafe.Pointer(form.Handle())
-	form.Show()
-	form.OnClose().Bind(func(arg *gform.EventArg) {
-		w32.DestroyWindow(w32.HWND(form.Handle()))
+	runtime.LockOSThread()
+
+	const className = "fullscreen_window_class"
+	classNamePtr, _ := syscall.UTF16PtrFromString(className)
+	w32.RegisterClassEx(&w32.WNDCLASSEX{
+		Cursor: w32.LoadCursor(0, w32.MakeIntResource(w32.IDC_ARROW)),
+		WndProc: syscall.NewCallback(func(window w32.HWND, msg uint32, w, l uintptr) uintptr {
+			switch msg {
+			case w32.WM_KEYDOWN:
+				if w == w32.VK_ESCAPE {
+					w32.SendMessage(window, w32.WM_CLOSE, 0, 0)
+				}
+				return 0
+			case w32.WM_DESTROY:
+				w32.PostQuitMessage(0)
+				return 0
+			default:
+				return w32.DefWindowProc(window, msg, w, l)
+			}
+		}),
+		ClassName: classNamePtr,
 	})
+
+	windowNamePtr, _ := syscall.UTF16PtrFromString("Static Triangle")
+	windowHandle := w32.CreateWindow(
+		classNamePtr,
+		windowNamePtr,
+		w32.WS_OVERLAPPEDWINDOW|w32.WS_VISIBLE,
+		w32.CW_USEDEFAULT, w32.CW_USEDEFAULT, 640, 480,
+		0, 0, 0, nil,
+	)
 
 	d3d, err := d3d9.Create(d3d9.SDK_VERSION)
 	check(err)
@@ -77,16 +95,17 @@ func main() {
 	check(device.SetVertexDeclaration(decl))
 
 	// create a timer that ticks every 100ms and register a callback for it
-	w32.SetTimer(w32.HWND(windowHandle), 1, 100, 0)
-	form.Bind(w32.WM_TIMER, func(*gform.EventArg) {
+	w32.SetTimer(windowHandle, 1, 100, 0)
+	var msg w32.MSG
+	for w32.GetMessage(&msg, 0, 0, 0) != 0 {
+		w32.TranslateMessage(&msg)
 		check(device.Clear(nil, d3d9.CLEAR_TARGET, 0, 0, 0))
 		check(device.BeginScene())
 		check(device.DrawPrimitive(d3d9.PT_TRIANGLELIST, 0, 1))
 		check(device.EndScene())
 		check(device.Present(nil, nil, 0, nil))
-	})
-
-	gform.RunMainLoop()
+		w32.DispatchMessage(&msg)
+	}
 }
 
 func check(err error) {

